rocket_learn/rollout_generator/redis/redis_rollout_worker.py

```diff
@@ -210,15 +220,18 @@ def _get_past_model(self, version):
         return model
 
     def select_gamemode(self):
-        mode_exp = {m.decode("utf-8"): int(v) for m, v in self.redis.hgetall(EXPERIENCE_PER_MODE).items()}
-        if self.gamemode_weights is None:
-            mode = min(mode_exp, key=mode_exp.get)
-        else:
-            total = sum(mode_exp.values()) + 1e-8
-            mode_exp = {k: mode_exp[k] / total for k in mode_exp.keys()}
-            # find exp which is farthest below desired exp
-            diff = {k: self.gamemode_weights[k] - mode_exp[k] for k in mode_exp.keys()}
-            mode = max(diff, key=diff.get)
+
+        total = sum(self.experience.values()) + 1e-8
+        mode_exp = {k: self.experience[k] / total for k in self.experience.keys()}
+        diff = {k: self.target_weights[k] - mode_exp[k] for k in mode_exp.keys()}
+        # change diff from experience weights to gamemode weights
+        for k in diff.keys():
+            b, o = k.split("v")
+            diff[k] *= int(b)
+        self.gamemode_weights = {k: max(self.gamemode_weights[k] + diff[k], 0) for k in self.gamemode_weights.keys()}
+        new_sum = sum(self.gamemode_weights.values())
+        self.gamemode_weights = {k: self.gamemode_weights[k] / new_sum for k in self.gamemode_weights.keys()}
+        mode = np.random.choice(list(self.gamemode_weights.keys()), p=list(self.gamemode_weights.values()))
```

Comment:
Although this method improves on the previous one in several ways (uses random sampling and reduces worker correlation), they still both suffer from the same downfall: The algorithm is based on a proportional correction based on the error between target and current distribution. 
This means that as the current distribution gets closer to the target one, the error goes to zero and so does the correction term, which causes the error to go back up.
Nevertheless, I can see how this new algorithm is more robust than the previous one when multiple workers are sampling in parallel but the empirical distribution will probably never reach the target one.